kms/uri/uri.go

```diff
@@ -0,0 +1,86 @@
+package uri
+
+import (
+	"net/url"
+	"strings"
+
+	"github.com/pkg/errors"
+)
+
+// URI implements a parser for a URI format based on the the PKCS #11 URI Scheme
+// defined in https://tools.ietf.org/html/rfc7512
+//
+// These URIs will be used to define the key names in a KMS.
+type URI struct {
+	*url.URL
+	Values url.Values
+}
+
+// New creates a new URI from a scheme and key-value pairs.
+func New(scheme string, values url.Values) *URI {
+	return &URI{
+		URL: &url.URL{
+			Scheme: scheme,
+			Opaque: strings.ReplaceAll(values.Encode(), "&", ";"),
+		},
+		Values: values,
+	}
+}
+
+// NewFile creates an uri for a file.
+func NewFile(path string) *URI {
+	return &URI{
+		URL: &url.URL{
+			Scheme: "file",
+			Path:   path,
+		},
+	}
+}
+
+// HasScheme returns true if the given uri has the given scheme, false otherwise.
+func HasScheme(scheme, rawuri string) bool {
+	u, err := url.Parse(rawuri)
+	if err != nil {
+		return false
+	}
+	return strings.EqualFold(u.Scheme, scheme)
+}
+
+// Parse returns the URI for the given string or an error.
+func Parse(rawuri string) (*URI, error) {
+	u, err := url.Parse(rawuri)
```

Comment:
It looks like 7512 allows pkcs11 uris to have query components, separated by `&`. So I'm not sure this will work generally but it probably works for what we're doing right now.

Reply:
`&` are used in the query string, and they will be part of the query. In any case, I've added a teste case in 6c9cd70.